pkg/descheduler/descheduler.go

```diff
@@ -358,74 +317,82 @@ func RunDeschedulerStrategies(ctx context.Context, rs *options.DeschedulerServer
-				if strategy.Enabled {
-					params := strategy.Params
-					if params == nil {
-						params = &api.StrategyParameters{}
-					}
-
-					nodeFit := false
-					if name != "PodLifeTime" {
-						nodeFit = params.NodeFit
-					}
-
-					// TODO(jchaloup): once all strategies are migrated move this check under
-					// the default evictor args validation
-					if params.ThresholdPriority != nil && params.ThresholdPriorityClassName != "" {
-						klog.V(1).ErrorS(fmt.Errorf("priority threshold misconfigured"), "only one of priorityThreshold fields can be set", "pluginName", name)
-						continue
-					}
-					thresholdPriority, err := utils.GetPriorityFromStrategyParams(ctx, rs.Client, strategy.Params)
-					if err != nil {
-						klog.ErrorS(err, "Failed to get threshold priority from strategy's params")
-						continue
-					}
+		var enabledDeschedulePlugins []framework.DeschedulePlugin
+		var enabledBalancePlugins []framework.BalancePlugin
 
-					defaultevictorArgs := &defaultevictor.DefaultEvictorArgs{
-						EvictLocalStoragePods:   evictLocalStoragePods,
-						EvictSystemCriticalPods: evictSystemCriticalPods,
-						IgnorePvcPods:           ignorePvcPods,
-						EvictFailedBarePods:     evictBarePods,
-						NodeFit:                 nodeFit,
-						PriorityThreshold: &api.PriorityThreshold{
-							Value: &thresholdPriority,
-						},
+		// Build plugins
+		for _, profile := range deschedulerPolicy.Profiles {
+			pc := getPluginConfig(defaultevictor.PluginName, profile.PluginConfigs)
+			if pc == nil {
+				klog.ErrorS(fmt.Errorf("unable to get plugin config"), "skipping plugin", "plugin", defaultevictor.PluginName)
+				continue
+			}
+			evictorFilter, err := defaultevictor.New(
+				pc.Args,
+				&handleImpl{
+					clientSet:                 rs.Client,
+					getPodsAssignedToNodeFunc: getPodsAssignedToNode,
+					sharedInformerFactory:     sharedInformerFactory,
+				},
+			)
+			if err != nil {
+				klog.ErrorS(fmt.Errorf("unable to construct a plugin"), "skipping plugin", "plugin", defaultevictor.PluginName)
+				continue
+			}
+			handle := &handleImpl{
+				clientSet:                 rs.Client,
+				getPodsAssignedToNodeFunc: getPodsAssignedToNode,
+				sharedInformerFactory:     sharedInformerFactory,
+				evictor: &evictorImpl{
+					podEvictor:    podEvictor,
+					evictorFilter: evictorFilter.(framework.EvictorPlugin),
+				},
+			}
+			// Assuming only a list of enabled extension points.
+			// Later, when a default list of plugins and their extension points is established,
+			// compute the list of enabled extension points as (DefaultEnabled + Enabled - Disabled)
+			for _, plugin := range append(profile.Plugins.Deschedule.Enabled, profile.Plugins.Balance.Enabled...) {
+				pc := getPluginConfig(plugin, profile.PluginConfigs)
+				if pc == nil {
+					klog.ErrorS(fmt.Errorf("unable to get plugin config"), "skipping plugin", "plugin", plugin)
+					continue
+				}
+				pg := pluginsMap[plugin](pc.Args, handle)
```

Comment:
should we guard against invalid plugin names in a profile?

```suggestion
				pgFunc, ok := pluginsMap[plugin]
				if !ok {
					klog.ErrorS(fmt.Errorf("invalid plugin name in profile"), "skipping plugin", "plugin", plugin, "profile", profile.Name)
					continue
				}
				pg := pgFunc(pc.Args, handle)
```

Reply:
Also, should we explicitly set the log level?
`klog.V(1).ErrorS`

Reply:
Actually, we should call the ErrorS without any log level since all errors are expected to be seen right away. I wonder how did this happen :) It got introduced by accident and copy-pasted all around.

Reply:
> should we guard against invalid plugin names in a profile?

Added the check